src/pds/aipgen/registry.py

```diff
@@ -0,0 +1,517 @@
+    bac[lidvid] = files                                               # Stash for future use
+
+
+def _comprehendRegistry(
+    url: str, bundleLIDVID: str, allCollections=True,
+    workaroundBadDocuments=True, workaroundPaginationBug=True
+) -> (int, dict, str):
+    '''Query the PDS API at ``url`` for all information about the PDS ``bundleLIDVID`` and return a
+    comprehension of it. If ``allCollections`` is True, we include every reference from a collection
+    that's LID-only; if it's False, then we only include the latest reference form a LID-only reference.
+    A "comprehension of it" means a triple of the common prefix length of all PDS paths referenced
+    within it, the "B.A.C." (a dict mapping PDS lidvids to sets of ``_File``s), and the title of
+    the PDS bundle.
+
+    Note: currently ``allCollections`` is ignored; see NASA-PDS/pds-api#74.
+    '''
+    _logger.debug('🤔 Comprehending the registry at %s for %s', url, bundleLIDVID)
+
+    # Set up our client connection
+    config = pds.api_client.Configuration()
+    config.host = url
+    apiClient = pds.api_client.ApiClient(config)
+
+    # This is the "B.A.C." 😏
+    bac = {}
+
+    bundle = _getBundle(apiClient, bundleLIDVID)  # There's no class "Bundle" but class Product 🤷‍♀️
+    if bundle is None:
+        raise ValueError(f'🤷‍♀️ The bundle {bundleLIDVID} cannot be found in the registry at {url}')
+    title = bundle.title if bundle.title else '«unknown»'
+    _addFiles(bundle, bac)
+
+    bundleURL = bundle.metadata.label_url
+    prefixLen = bundleURL.rfind('/') + 1
+
+    # It turns out the PDS registry make this *trivial* compared to the PDS filesystem version;
+    # Just understanding it all was there was the hard part! 😊 THANK YOU! 🙏
+    for collection in _getCollections(apiClient, bundleLIDVID, workaroundPaginationBug):
+        _addFiles(collection, bac)
+        for product in _getProducts(apiClient, collection.id, workaroundBadDocuments):
+            _addFiles(product, bac)
+
+    # C'est tout 🌊
+    return prefixLen, bac, title
+
+
+def _writeChecksumManifest(fn: str, prefixLen: int, bac: dict) -> (str, int, int):
+    '''Write an AIP "checksum manifest" to the given ``fn`` PDS filename, stripping ``prefixLen``
+    characters off paths, and using information from the ``bac``.  Return a triple of the MD5
+    of the manifest, its size in bytes, and a count of the number of entries in it.
+    '''
+    hashish, size, count = hashlib.new('md5'), 0, 0
+    with open(fn, 'wb') as o:
+        for files in bac.values():
+            for f in files:
+                entry = f'{f.md5}\t{f.url[prefixLen:]}\r\n'.encode('utf-8')
+                o.write(entry)
+                hashish.update(entry)
+                size += len(entry)
+                count += 1
+                if count % _progressLogging == 0:
+                    _logger.debug('⏲ Wrote %d entries into the checksum manifest %s', count, fn)
+    _logger.info('📄 Wrote AIP checksum manifest: %s', fn)
```

Comment:
Not sure how useful this is, but you could display the final `count` value here as well as a sanity check, since you'd only ever see increments of `_progressLogging` from the debug statement above (and even less if `count` winds up less than `_progressLogging`!).

Same goes for `_writeTransferManifest` below.

Reply:
There are 3 such loops, 2 of which use the same criteria (iteration over lidvid → set of files) but one which is just different (set of files). If the 2 become 3, I'll refactor, but for now I'm okay to live with it as is.